sympy/simplify/cse_diff.py

```diff
@@ -0,0 +1,190 @@
+        replaced_subexpr = node.xreplace(expr_to_replacement_cache)
+        replacement_to_reduced_expr_cache[replacement_symbol] = replaced_subexpr
+        expr_to_replacement_cache[node] = replacement_symbol
+        return replacement_symbol, replaced_subexpr
+
+    if not isinstance(expr, ImmutableDenseMatrix):
+        msg = (
+            'The forward Jacobian differentiation algorithm can only be used '
+            'to differentiate a single matrix expression at a time.'
+        )
+        raise NotImplementedError(msg)
+    elif expr.shape[1] != 1:
+        msg = 'Can only compute the Jacobian for column matrices.'
+        raise NotImplementedError(msg)
+    elif not isinstance(wrt, ImmutableDenseMatrix) or wrt.shape[1] != 1:
+        msg = (
+            'The forward Jacobian differentiation algorithm can only compute '
+            'Jacobians with respect to column matrices.'
+        )
+        raise NotImplementedError(msg)
+
+    symbols = expr.free_symbols
+    replacement_symbols = numbered_symbols(
+        prefix='_z',
+        cls=Symbol,
+        exclude=expr.free_symbols,
+    )
+
+    expr_to_replacement_cache = {}
+    replacement_to_reduced_expr_cache = {}
+
+    replacements, reduced_exprs = cse(expr.args[2], replacement_symbols)
+    for replacement_symbol, reduced_subexpr in replacements:
+        replaced_subexpr = reduced_subexpr.xreplace(expr_to_replacement_cache)
+        replacement_to_reduced_expr_cache[replacement_symbol] = replaced_subexpr
+        expr_to_replacement_cache[reduced_subexpr] = replacement_symbol
+        for node in postorder_traversal(reduced_subexpr):
+            _ = add_to_cache(node)
+    for reduced_expr in reduced_exprs:
+        for node in reduced_expr:
+            _ = add_to_cache(node)
+
+    reduced_matrix = ImmutableDenseMatrix(reduced_exprs).xreplace(expr_to_replacement_cache)
+    replacements = list(replacement_to_reduced_expr_cache.items())
+
+    absolute_derivative_mapping = {}
+    for i, wrt_symbol in enumerate(wrt.args[2]):
+        absolute_derivative = [S.Zero] * len(wrt)
+        absolute_derivative[i] = S.One
+        absolute_derivative_mapping[wrt_symbol] = ImmutableDenseMatrix([absolute_derivative])
+
+    zeros = ImmutableDenseMatrix.zeros(1, len(wrt))
+    for symbol, subexpr in replacements:
+        free_symbols = subexpr.free_symbols
+        absolute_derivative = zeros
+        for free_symbol in free_symbols:
+            replacement_symbol, partial_derivative = add_to_cache(subexpr.diff(free_symbol))
+            absolute_derivative += partial_derivative * absolute_derivative_mapping.get(free_symbol, zeros)
+
+        # Modification to manage dynamicsymbols
+        if free_symbols == {dynamicsymbols._t}:
+            absolute_derivative_mapping[symbol] = ImmutableDenseMatrix([[subexpr.diff(sub) for sub in wrt]])
+            continue
+        absolute_derivative_mapping[symbol] = ImmutableDenseMatrix([[add_to_cache(a)[0] for a in absolute_derivative]])
```

Comment:
Why is this special case needed?

It would be better not to have a special case for dynamicsymbols.

Reply:
Yes, this should simply work with functions. The concept of "dynamic symbols" should remain in the physics vector package.

Reply:
@moorepants Unfortunately this is the only way found to make it work with expressions containing dynamicsymbols for now.  But indeed if the function is placed in cse_diff it should not contain this special case.

Reply:
A "dynamic symbol" is simply a `Function('x')(Symbol('y'))`. If this function stays in this location, then we need to make it general and handle any SymPy function, which is a super set of `Function('x')(dynamicsymbols._t)`.